lib/propolis/src/hw/virtio/softnpu.rs

```diff
@@ -0,0 +1,1061 @@
+            let pl = match &mut *p {
+                Some(ref mut pl) => &mut pl.1,
+                None => continue, // no program is loaded
+            };
+
+            Self::process_external_packet(
+                index + 1,
+                pkt,
+                &data_handles,
+                virtio.clone(),
+                pl,
+                &log,
+            )
+        }
+    }
+
+    /// Run a packet coming into the ASIC from an external port through the
+    /// loaded pipeline and forward it on to its destination.
+    fn process_external_packet<'a>(
+        index: usize,
+        mut pkt: packet_in<'a>,
+        data_handles: &Vec<dlpi::DlpiHandle>,
+        virtio: Arc<PortVirtioState>,
+        pipeline: &mut Box<dyn Pipeline>,
+        log: &Logger,
+    ) {
+        for (mut out_pkt, port) in
+            pipeline.process_packet(index as u16, &mut pkt)
+        {
+            // packet is going to CPU port
+            if port == 0 {
+                Self::send_packet_to_cpu_port(&mut out_pkt, &virtio, &log);
+            }
+            // packet is passing through
+            else {
+                Self::send_packet_to_ext_port(
+                    &mut out_pkt,
+                    data_handles,
+                    port - 1,
+                    &log,
+                );
+            }
+        }
+    }
+
+    /// Run a packet coming into the ASIC from the guest pci port through the
+    /// loaded pipeline and forward it on to its destination.
+    fn process_guest_packet<'a>(
+        mut pkt: packet_in<'a>,
+        data_handles: &Vec<dlpi::DlpiHandle>,
+        pipeline: &mut Box<dyn Pipeline>,
+        log: &Logger,
+    ) {
+        for (mut out_pkt, port) in pipeline.process_packet(0, &mut pkt) {
+            if port == 0 {
+                // no looping packets back to the guest
+                return;
+            }
+            Self::send_packet_to_ext_port(
+                &mut out_pkt,
+                data_handles,
+                port - 1,
+                &log,
+            );
+        }
+    }
+
+    /// Send a packet out an external port using dlpi.
+    fn send_packet_to_ext_port<'a>(
+        pkt: &mut packet_out<'a>,
+        data_handles: &Vec<dlpi::DlpiHandle>,
+        port: u16,
+        log: &Logger,
+    ) {
+        // get the dlpi handle for this port
+        let dh = data_handles[port as usize];
+
+        //TODO avoid copying the whole packet
+        let mut out = pkt.header_data.clone();
+        out.extend_from_slice(pkt.payload_data);
+
+        if let Err(e) = dlpi::send(dh, &[], out.as_slice(), None) {
+            error!(log, "tx (ext,0): {}", e);
+        }
+    }
+
+    /// Send a packet out the guest pci port using virtio.
+    fn send_packet_to_cpu_port<'a>(
+        pkt: &mut packet_out<'a>,
+        virtio: &Arc<PortVirtioState>,
+        _log: &Logger,
+    ) {
+        let mem = virtio.pci_state.acc_mem.access().unwrap();
```

Comment:
We should probably be more graceful here if we cannot access guest memory through the pci state handle.  Such would be the case if the guest happened to reset the device and/or disable bus-mastering.  In that case, this handler can assume there is no free descriptor into which the packet can be placed.

Reply:
Now checking for none, kicking out a warning and returning.